Add a function to take a view of an index set with regard to an index set mask.

bangerth opened the pull request:
I will need this for https://github.com/geodynamics/aspect/issues/5412, but I think it will, at some point or other, be of independent interest as well.

kronbichler commented on source/base/index_set.cc:
```diff
@@ -257,6 +257,144 @@ IndexSet::get_view(const size_type begin, const size_type end) const
+
+  // Now turn the ranges of overlap we have accumulated into an IndexSet in
+  // its own right:
+  IndexSet result(size());
```
Didn't you want to make the size equal the number of entries in the mask: The documentation states `then the result will be an index set for a space of size 50 that contains those indices that [...]`. Overall, I am not sure yet how you want to achieve this in a parallel setting. If I read the documentation correctly, I need to be sure that the mask can be defined for the entirety of the indices on **all processes** in the user code, and only afterwards call this here. Is this understanding correct?

bangerth replied:
Oh yes, excellent catch! This should have been 
```suggestion
  IndexSet result(mask.n_elements());
```
I made the fix, and also added assertions in this regard to the three tests added here.

bangerth commented:
Thanks for the careful reading, @kronbichler . I fixed the bug, and I've also added an example to the documentation that explains the sort of context in which I want to use this function.
